### PR TITLE
Add Subscription Topic Handler

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Altair Work

**What does this PR do? Why is it needed?**

- [x] Add in subscription topic handler to manage multiple gossip topics a node is susbcribed to.
- [x] Add in tests for it.

**Which issues(s) does this PR fix?**

#8638 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/sync/subscription_topic_handler.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		log.WithError(err).Error("Could not retrieve digest")
+		return
+	}
+	s.digestMap[digest] += 1
+}
+
+func (s *subTopicHandler) topicExists(topic string) bool {
+	s.RLock()
+	_, ok := s.subTopics[topic]
+	s.RUnlock()
```
Comment (jmozah) [inline beacon-chain/sync/subscription_topic_handler.go:41]: defer this as soon as you lock.

File: beacon-chain/sync/subscription_topic_handler.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	return ok
+}
+
+func (s *subTopicHandler) removeTopic(topic string) {
+	s.Lock()
+	defer s.Unlock()
+	delete(s.subTopics, topic)
```
Comment (jmozah) [inline beacon-chain/sync/subscription_topic_handler.go:48]: better to  delete this after the digest is decremented succesfully.

**Review by jmozah — COMMENTED** [review]
LGTM except some small issues.

File: beacon-chain/sync/subscription_topic_handler.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/sync/subscription_topic_handler.go:48]: Slightly trickier than that as the 2 maps are technically unrelated, 
`subTopics` tracks active subscriptions and
`digestMap` tracks the number of topics registered for a particular digest. 

Even if `digestMap` doesn't have the appropriate digest mapped(should not be possible), we should still remove it from the subscription map. Which is why i remove it first.

File: beacon-chain/sync/subscription_topic_handler.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		digestMap: map[[4]byte]int{},
+	}
+}
+
+func (s *subTopicHandler) addTopic(topic string, sub *pubsub.Subscription) {
+	s.Lock()
+	defer s.Unlock()
+	s.subTopics[topic] = sub
+	digest, err := p2p.ExtractGossipDigest(topic)
+	if err != nil {
```
Comment (terencechain) [inline beacon-chain/sync/subscription_topic_handler.go:31]: Potential race condition. if this errs out, we will be stuck with `s.subTopics[topic] == sub` but no update to `digestMap`

Comment (terencechain) [reply beacon-chain/sync/subscription_topic_handler.go:31]: I see your following comment @jmozah. Sounds like it's intentional, even `digestMap` fails we should still update the subscription map

**Review by terencechain — APPROVED** [review]

**Pull request merged**